62.CAD/main.cpp

```diff
@@ -364,61 +387,88 @@ class CPolyline
 			you initially set them to 0 in addLinePoints/addQuadBeziers
 
 		NOTE that PolylineConnectors are special object types, user does not add them and they should not be part of m_sections vector
-	*/ 
+	*/
+
+	void preprocessPolylineWithStyle(const CPULineStyle& lineStyle)
+	{
+		float phaseShiftTotal = lineStyle.phaseShift;
+		for (uint32_t i = 0u; i < m_sections.size(); i++)
+		{
+			const auto& section = m_sections[i];
+
+			if (section.type == ObjectType::LINE)
+			{
+				// calculate phase shift at each point of each line in section
+				const uint32_t linePointCnt = section.count + 1u;
+				for (uint32_t i = 0u; i < linePointCnt; i++)
+				{
+					const uint32_t currIdx = section.index + i;
+					auto& linePoint = m_linePoints[currIdx];
+					if (i == 0u)
+					{
+						linePoint.phaseShift = phaseShiftTotal;
+						continue;
+					}
+
+					const auto& prevLinePoint = m_linePoints[section.index + i - 1u];
+					const double lineLen = glm::length(linePoint.p - prevLinePoint.p);
+
+					const double changeInPhaseShiftBetweenCurrAndPrevPoint = std::remainder(lineLen, 1.0f / lineStyle.reciprocalStipplePatternLen) * lineStyle.reciprocalStipplePatternLen;
+					linePoint.phaseShift = glm::fract(phaseShiftTotal + changeInPhaseShiftBetweenCurrAndPrevPoint); // TODO: fract may be unecessary since we fract anyway in fragment shader
+					phaseShiftTotal = linePoint.phaseShift;
+				}
+			}
+			else if (section.type == ObjectType::QUAD_BEZIER)
+			{
+				// calculate phase shift at point P0 of each bezier
+				const uint32_t quadBezierCnt = section.count;
+				for (uint32_t i = 0u; i < quadBezierCnt; i++)
+				{
+					const uint32_t currIdx = section.index + i;
+					QuadraticBezierInfo& quadBezierInfo = m_quadBeziers[currIdx];
+					if (i == 0u)
+					{
+						quadBezierInfo.phaseShift = phaseShiftTotal;
+						continue;
+					}
+
+					const QuadraticBezierInfo& prevQuadBezierInfo = m_quadBeziers[currIdx - 1u];
+					shapes::QuadraticBezier<double> quadraticBezier = shapes::QuadraticBezier<double>::construct(prevQuadBezierInfo.p[0], prevQuadBezierInfo.p[1], prevQuadBezierInfo.p[2]);
+					shapes::Quadratic<double> quadratic = shapes::Quadratic<double>::constructFromBezier(quadraticBezier);
+					shapes::Quadratic<double>::ArcLengthCalculator arcLenCalc = shapes::Quadratic<double>::ArcLengthCalculator::construct(quadratic);
+					const double lineLen = arcLenCalc.calcArcLen(1.0f);
+
+					const double nextLineInSectionLocalPhaseShift = std::remainder(lineLen, 1.0f / lineStyle.reciprocalStipplePatternLen) * lineStyle.reciprocalStipplePatternLen;
+					quadBezierInfo.phaseShift = glm::fract(phaseShiftTotal + nextLineInSectionLocalPhaseShift); // TODO: fract may be unecessary since we fract anyway in fragment shader
+					phaseShiftTotal = quadBezierInfo.phaseShift;
+				}
+
+				// calculate phase shift at end of the section, unecessary if last section
+				if(i < m_sections.size() - 1u)
```

Comment:
you made this unnecessarily complex. you know you can just extend the previous for loop right?

you're always doing `phaseShiftTotal = glm::fract(phaseShiftTotal + additionalPhaseShift);

just do
```c++
double additionalPhaseShift=0.0;
if (i > 0)
{
// calculate additionalPhaseShift based on previous bezier arclen
}
// do  `phaseShiftTotal = glm::fract(phaseShiftTotal + additionalPhaseShift);`
if (i < quadBezierCnt)
{
m_quadBeziers[currIdx].phaseShift=totalPhaseShift;
}
```